File: src/io/fdeitylink/keroedit/script/ScriptEditTab.java
package io.fdeitylink.keroedit.script;

import java.text.MessageFormat;

import java.util.stream.Stream;
import java.util.stream.Collectors;

import java.nio.charset.Charset;

import java.nio.file.Files;
import java.nio.file.Path;

import java.io.IOException;
import java.io.FileNotFoundException;

import javafx.scene.control.Alert;

import javafx.scene.control.TextArea;
import javafx.scene.text.Font;


import io.fdeitylink.util.NullArgumentException;

import io.fdeitylink.util.UtilsKt;

import io.fdeitylink.util.fx.FileEditTab;

import io.fdeitylink.keroedit.Messages;

import io.fdeitylink.util.fx.FXUtil;

import io.fdeitylink.keroedit.gamedata.GameData;

import io.fdeitylink.keroedit.mapedit.MapEditTab;

public final class ScriptEditTab extends FileEditTab {
    private final TextArea textArea;

    public ScriptEditTab(Path inPath) throws IOException {
        //TODO: Check if inPath is actually within the mod folder and throw except if not?
        super(inPath = inPath.toAbsolutePath());

        /*
         * TODO: Fix this weird exception that I can't seem to get myself
         * java.io.UncheckedIOException: java.nio.charset.MalformedInputException: Input length = 1
	     * java.io.BufferedReader$1.hasNext(Unknown Source)
	     * java.util.Iterator.forEachRemaining(Unknown Source)
         * java.util.Spliterators$IteratorSpliterator.forEachRemaining(Unknown Source)
         * java.util.stream.AbstractPipeline.copyInto(Unknown Source)
         * java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
         * java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)
         * java.util.stream.AbstractPipeline.evaluate(Unknown Source)
         * java.util.stream.ReferencePipeline.collect(Unknown Source)
         * io.fdeitylink.keroedit.script.ScriptEditTab.<init>(ScriptEditTab.java:44)
         *  - scriptText = lineStream.collect(Collectors.joining("\n"));
         */

        String scriptText = "";
        try (Stream <String> lines = Files.lines(inPath, Charset.forName("Shift_JIS"))) {
            scriptText = lines.collect(Collectors.joining("\n"));
        }
        catch (final FileNotFoundException except) {
            try {
                Files.createFile(inPath);
            }
            catch (final IOException ex) {
                //TODO: Do nothing or throw except?
            }
        }
        catch (final IOException except) {
            FXUtil.createAlert(Alert.AlertType.ERROR, Messages.getString("ScriptEditTab.IOExcept.TITLE"), null,
                               MessageFormat.format(Messages.getString("ScriptEditTab.IOExcept.MESSAGE"),
                                                    inPath.getFileName(), except.getMessage())).showAndWait();
            throw except;
        }

        textArea = new TextArea(scriptText);
        textArea.requestFocus();
        textArea.setFont(new Font("Consolas", 12));

        /*
         * Assume that this ScriptEditTab is not inside of a
         * MapEditTab and set the title to the filename. The
         * other constructor will call setText() again if this
         * tab is inside of a MapEditTab.
         */
        setText(UtilsKt.baseFilename(inPath, GameData.scriptExtension));

        setContent(textArea);
    }

    public ScriptEditTab(final Path inPath, final MapEditTab parent) throws IOException {
        this(inPath); //throws NullArgumentException if inPath == null

        NullArgumentException.Companion.requireNonNull(parent, "ScriptEditTab", "parent");
        textArea.textProperty().addListener((observable, oldValue, newValue) -> {
            markChanged();
            parent.markChanged();
        });

        setText(Messages.getString("ScriptEditTab.TITLE"));
    }

    @Override
    public void undo() {
        textArea.undo();
    }

    @Override
    public void redo() {
        textArea.redo();
    }

    @Override
    public void save() {
        final Path scriptPath = getPath();
        try {
            Files.write(scriptPath, textArea.getParagraphs(), Charset.forName("Shift_JIS"));
            markUnchanged();
        }
        catch (final IOException except) {
            FXUtil.createAlert(Alert.AlertType.ERROR, Messages.getString("ScriptEditTab.Save.IOExcept.TITLE"), null,
                               MessageFormat.format(Messages.getString("ScriptEditTab.Save.IOExcept.MESSAGE"),
                                                    scriptPath.getFileName(), except.getMessage())).showAndWait();
        }
    }

    //Made public so the parent MapEditTab can call it
    @Override
    public void markUnchanged() {
        super.markUnchanged();
    }
}

File: src/io/fdeitylink/keroedit/script/ScriptEditTab.kt
package io.fdeitylink.keroedit.script

import java.text.MessageFormat

import java.nio.file.Path
import java.nio.file.Files

import java.io.IOException

import java.nio.charset.Charset

import java.util.stream.Collectors

import javafx.scene.control.Alert

import javafx.scene.control.TextArea
import javafx.scene.text.Font

import io.fdeitylink.util.baseFilename

import io.fdeitylink.util.use

import io.fdeitylink.util.fx.FXUtil
import io.fdeitylink.util.fx.FileEditTab

import io.fdeitylink.keroedit.Messages

import io.fdeitylink.keroedit.gamedata.GameData

import io.fdeitylink.keroedit.mapedit.MapEditTab
import io.fdeitylink.util.Logger

class ScriptEditTab
@Throws(IOException::class) constructor(inPath: Path,
                                        private val parent: MapEditTab? = null
) : FileEditTab(inPath.toAbsolutePath().apply {
    if (!Files.exists(this)) {
        try {
            Files.createFile(this)
        }
        catch (except: IOException) {
            Logger.logThrowable("Error creating new script file $this", except)
        }
    }
}) {

    private val textArea = TextArea()

    init {
        /*
         * TODO: Fix this weird exception that I can't seem to get myself
         * java.io.UncheckedIOException: java.nio.charset.MalformedInputException: Input length = 1
	     * java.io.BufferedReader$1.hasNext(Unknown Source)
	     * java.util.Iterator.forEachRemaining(Unknown Source)
         * java.util.Spliterators$IteratorSpliterator.forEachRemaining(Unknown Source)
         * java.util.stream.AbstractPipeline.copyInto(Unknown Source)
         * java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
         * java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)
         * java.util.stream.AbstractPipeline.evaluate(Unknown Source)
         * java.util.stream.ReferencePipeline.collect(Unknown Source)
         * io.fdeitylink.keroedit.script.ScriptEditTab.<init>(ScriptEditTab.java:44)
         *  - scriptText = lineStream.collect(Collectors.joining("\n"));
         */

        val scriptText = try {
            Files.lines(path, Charset.forName("Shift_JIS")).use {
                //TODO: Replace all newlines with \n? (use regex)
                it.collect(Collectors.joining("\n"))
            }
        }
        catch (except: IOException) {
            FXUtil.createAlert(type = Alert.AlertType.ERROR,
                               title = Messages["ScriptEditTab.IOExcept.TITLE"],
                               message = MessageFormat.format(Messages["Keroedit.IOExcept.MESSAGE"],
                                                              path.fileName, except.message)).showAndWait()
            //TODO: Is this necessary?
            tabPaneProperty().addListener { _, _, _ -> tabPane?.tabs?.remove(this) }
            throw except
        }

        textArea.text = scriptText
        textArea.requestFocus()
        textArea.font = Font.font("Consolas", 12.0)
        textArea.textProperty().addListener { _, _, _ -> markChanged() }

        text = if (null == parent) path.baseFilename(GameData.scriptExtension) else Messages["ScriptEditTab.TITLE"]

        content = textArea
    }

    override fun undo() = textArea.undo()

    override fun redo() = textArea.redo()

    override fun save() {
        try {
            Files.write(path, textArea.paragraphs, Charset.forName("Shift_JIS"))
            markUnchanged()
        }
        catch (except: IOException) {
            FXUtil.createAlert(type = Alert.AlertType.ERROR, title = Messages["ScriptEditTab.Save.IOExcept.TITLE"],
                               message = MessageFormat.format(Messages["ScriptEditTab.Save.IOExcept.MESSAGE"],
                                                              path.fileName, except.message)).showAndWait()
        }
    }

    override fun markChanged() {
        super.markChanged()
        parent?.markChanged()
    }

    override public fun markUnchanged() = super.markUnchanged()
}